### PR TITLE
Support MXNetTrainer

### PR DESCRIPTION
**Pull request by hkvision**
TODO: Add doc, readme, integration test.

### CODE REVIEW

**Comment by hkvision** [comment]
@jason-dai An initial PR with LeNet example is ready. I would merge it first. Subsequent PRs for polishing this will come very soon.

**Comment by hkvision** [comment]
Manually tested the example on yarn with built whl.

**Comment by hkvision** [comment]
Added unit tests for gluon and symbol. This would be enough for an initial merge.

File: pyzoo/zoo/ray/mxnet/mxnet_runner.py [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+        self.num_workers = config["num_workers"]
+        self.num_servers = config["num_servers"] if "num_servers" in self.config \
+            else self.num_workers
+
+        # Generate actor class
+        # Add a dummy custom resource for server to diff from worker
+        Worker = ray.remote(num_cpus=runner_cores, resources={"_mxnet_worker": 1})(MXNetRunner) \
+            if runner_cores else ray.remote(MXNetRunner)
+        Server = ray.remote(num_cpus=runner_cores, resources={"_mxnet_server": 1})(MXNetRunner) \
+            if runner_cores else ray.remote(MXNetRunner)
```
Comment (jason-dai) [inline pyzoo/zoo/ray/mxnet/mxnet_runner.py:223]: Is "_mxnet_worker" or "_mxnet_server" used anywhere?

File: pyzoo/zoo/ray/mxnet/mxnet_runner.py [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+
+        # Start runners: workers followed by serers
```
Comment (jason-dai) [inline pyzoo/zoo/ray/mxnet/mxnet_runner.py:225]: followed by servers

File: pyzoo/zoo/ray/mxnet/mxnet_runner.py [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+            for i in range(self.num_servers)
+        ]
+
+        # Compute URL for initializing distributed setup
+        ips = ray.get(
+            [runner.get_node_ip.remote() for runner in self.runners])
+        ports = ray.get(
+            [runner.find_free_port.remote() for runner in self.runners])
+        print(ips)
+        print(ports)
```
Comment (jason-dai) [inline pyzoo/zoo/ray/mxnet/mxnet_runner.py:241]: remove print

File: pyzoo/zoo/ray/mxnet/mxnet_runner.py [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+        if self.is_worker:
+            os.environ.update(env)
+            self.kv = mx.kv.create("dist_sync")
+            # Set seed so that the model on each worker is initialized with the same weights
+            if "seed" in self.config:
+                mx.random.seed(self.config["seed"])
+            data = self.data_creator(self.config, self.kv)
+            if isinstance(data, tuple):
+                assert len(data) == 1 or len(data) == 2, \
+                    "data_creator should return train data (and val data)"
```
Comment (jason-dai) [inline pyzoo/zoo/ray/mxnet/mxnet_runner.py:55]: need clearer error messages

File: pyzoo/zoo/ray/mxnet/mxnet_runner.py [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+                 # Users can specify train_function (not recommended) or use the default one.
+                 train_function=None,
+                 # Specify cpu resources for actors if necessary.
+                 runner_cores=None):
+        self.config = config
+        self.data_creator = data_creator
+        self.model_creator = model_creator
+        self.loss_creator = loss_creator
+        self.metrics_creator = metrics_creator
+        self.train_function = train_function
```
Comment (jason-dai) [inline pyzoo/zoo/ray/mxnet/mxnet_runner.py:213]: I think we should just remove `train_function` for now

File: pyzoo/test/zoo/ray/test_mxnet_symbol.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        config = {
+            "num_workers": 2,
+            "batch_size": 32,
+            "optimizer": "sgd",
+            "init": mx.init.Xavier(),
+            "optimizer_params": {'learning_rate': 0.01},
+            "log_interval": 2,
+            "seed": 42
+        }
+        trainer = MXNetTrainer(config, get_data_iters, get_model, metrics_creator=get_metrics)
```
Comment (jason-dai) [inline pyzoo/test/zoo/ray/test_mxnet_symbol.py:73]: no loss needed for symbolic model?

File: pyzoo/zoo/ray/mxnet/mxnet_runner.py [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+                            batch.data[0].astype("float32"), ctx_list=[mx.cpu()], batch_axis=0)
+                        label = gluon.utils.split_and_load(
+                            batch.label[0].astype("float32"), ctx_list=[mx.cpu()], batch_axis=0)
+                        outputs = []
+                        Ls = []
+                        from mxnet import autograd as ag
+                        with ag.record():
+                            for x, y in zip(data, label):
+                                z = self.model(x)  # forward
+                                L = self.loss(z, y)
```
Comment (jason-dai) [inline pyzoo/zoo/ray/mxnet/mxnet_runner.py:112]: what if loss is None?

File: pyzoo/test/zoo/ray/test_mxnet_symbol.py (hunk 6 above)
Comment (hkvision) [reply pyzoo/test/zoo/ray/test_mxnet_symbol.py:73]: Yes. Loss is not needed, it is included in the model (I think somehow similar to definition of tensorflow graph)
For example, http://beta.mxnet.io/r/api/mx.symbol.SoftmaxOutput.html SoftmaxOutput includes cross entropy.

File: pyzoo/zoo/ray/mxnet/mxnet_runner.py (hunk 1 above)
Comment (hkvision) [reply pyzoo/zoo/ray/mxnet/mxnet_runner.py:223]: This should be added as custom resource when we init ray, which currently needs to be done by users when creating RayContext.
Any better way to handle this in our side? I'm thinking of that.

File: pyzoo/zoo/ray/mxnet/mxnet_runner.py (hunk 3 above)
Comment (hkvision) [reply pyzoo/zoo/ray/mxnet/mxnet_runner.py:241]: Change to logging? Sometimes this would be useful to immediately find out what nodes the program is running on?

File: pyzoo/zoo/ray/mxnet/mxnet_runner.py (hunk 4 above)
Comment (hkvision) [reply pyzoo/zoo/ray/mxnet/mxnet_runner.py]: Updated.

File: pyzoo/zoo/ray/mxnet/mxnet_runner.py (hunk 5 above)
Comment (hkvision) [reply pyzoo/zoo/ray/mxnet/mxnet_runner.py]: Removed.

File: pyzoo/zoo/ray/mxnet/mxnet_runner.py (hunk 7 above)
Comment (hkvision) [reply pyzoo/zoo/ray/mxnet/mxnet_runner.py:110]: added assert loss not None if user inputs a gluon model.

File: pyzoo/zoo/ray/mxnet/mxnet_runner.py (hunk 1 above)
Comment (hkvision) [reply pyzoo/zoo/ray/mxnet/mxnet_runner.py:220]: https://github.com/intel-analytics/analytics-zoo/blob/master/pyzoo/zoo/ray/util/raycontext.py#L96
Add _mxnet_worker and _mxnet_trainer here? This wouldn't affect other applications without MXNet.
Also is trainer and ps in the code necessary? (this is for tensorflow only??)
@jason-dai

Comment (jason-dai) [reply pyzoo/zoo/ray/mxnet/mxnet_runner.py:220]: Yes, I think we can add _mxnet_worker and _mxnet_trainer here if needed. I don't think anyone uses trainer and ps?

Comment (hkvision) [reply pyzoo/zoo/ray/mxnet/mxnet_runner.py:220]: Yeah, seems trainer and ps are not used anywhere and I removed them in our code first. If users want these tags, they can add by themselves in RayContext.

Comment (hkvision) [reply pyzoo/zoo/ray/mxnet/mxnet_runner.py:221]: Updated. Added to labels, the special resources for mxnet wouldn't affect other usages.

**Comment by hkvision** [comment]
http://10.239.47.210:18888/job/ZOO-PR-Python-Spark-2.4-py36/1570/
spark2.4.3 py36 passed

**Comment by hkvision** [comment]
Style check passed: http://10.239.47.210:18888/view/ZOO-PR/job/ZOO-PR-Scala-StyleCheck/3196/

**Comment by hkvision** [comment]
http://10.239.47.210:18888/view/ZOO-PR/job/ZOO-PR-Python-Spark-2.4-py37/4/
spark 2.4.3 py37 passed

**Comment by hkvision** [comment]
Merge it first. Subsequent PRs will polish this initial version.
